### PR TITLE
New Settings: action sheets for adding new devices

### PR DESCRIPTION
**Pull request by rickpasetto**
This PR implements the action sheets for choosing new devices to add to the new settings UI.
The tricky part about this was refactoring it so that code could be reused.  In particular, I took some liberties in refactoring the implementation of the use case for `PumpManager`s that can behave as `CGMManager`s.  If the configured `pumpManager` implements `CGMManager`, then it appears as an `AvailableDevice`, and can be set up as such.

@ps2 I'm especially interested in your thoughts on this approach.  There's a lot of mixed responsibilities that I am trying to sort of clean up...but I may have some assumptions wrong.  A live PR review might help...

### CODE REVIEW

**Review by novalegra — COMMENTED** [review]
I like this approach, though I'd defer to @ps2 as to if this would work as intended. If y'all end up doing a live PR review, I'd love to come as well to hear the feedback

File: Loop/Managers/DeviceDataManager.swift
Comment (ps2) [inline Loop/Managers/DeviceDataManager.swift]: nit: The verb of this method is 'setup', and I would expect it "setup" the CGMManager in some way. But it returns an object that indicates if setup is needed. Perhaps something like `getCGMSetupState(of identifier: String) -> CGMSetupState`

File: Loop/View Controllers/StatusTableViewController.swift
Comment (ps2) [inline Loop/View Controllers/StatusTableViewController.swift]: nit: This method isn't part of CGMManagerSetupViewControllerDelegate; including it in the extension declaring conformance with CGMManagerSetupViewControllerDelegate initially made me think that it was.

File: Loop/View Controllers/StatusTableViewController.swift
Comment (ps2) [inline Loop/View Controllers/StatusTableViewController.swift]: I'm wondering why we would be in the state of presenting the user with an "Add CGM" button, when the CGM is already setup?

**Review by ps2 — COMMENTED** [review]
Most of this looks good to me. I do have a question about the state space though. The code reads as if Loop can be in a state where a CGM is configured, but we're presenting an "Add CGM" button to them?

File: Loop/View Controllers/StatusTableViewController.swift
Comment (rickpasetto) [reply Loop/View Controllers/StatusTableViewController.swift]: This is a good point....probably a bad naming of this enum value, too (as well as this function).  Really the enum is for responding with an actual `CGMManager` instance, vs. a `CGMManager.Type` that needs configuring.  I feel like a different name might make this clearer, I'll try that.

**Comment by rickpasetto** [comment]
@ps2 I've renamed the functions and the enum values.  LMK what you think

File: Loop/Managers/DeviceDataManager.swift
Comment (ps2) [inline Loop/Managers/DeviceDataManager.swift]: This type is still somewhat confusing to me (mixing two seemingly separate ideas: 'presence' and 'setup state'), and ends up creating handling code that opens up new questions (the "Is this what we want to do?" comment).  There should never be a `noSuchCGM`. Perhaps in place of checkForCGMManager:
```
// Returns a CGMManagerUI type if cgm setup is needed
public func didSelectCGMManager(_ identifier: String) -> CGMManagerUI? {
    [assign cgmManager if already configured, or already implemented by pumpmanager, otherwise return CGMManagerUI type]
}
```
That would obviate the need for a new type, and simplify the number of possibilities a caller needs to consider.

**Review by ps2 — APPROVED** [review]
Approving with a suggestion: I still find it a little confusing, and suggested an alternative. I think what you have works, though, so if you don't find it to be an improvement, let's go with what you have.

File: Loop/Managers/DeviceDataManager.swift
Comment (rickpasetto) [reply Loop/Managers/DeviceDataManager.swift]: @ps2 I see what you mean.  I agree with moving he responsibility of _assigning_ the newly-minted `CGMManager` _to_ the `DeviceDataManager`.  After all, that's who needs it.  All we really are doing, if I understand this correctly, is providing the possibility of the caller needing to start a setup flow if need be, and then tell `DeviceDataManager` of the result.  Something like this:

```swift
/// This is what I _want_ to do
    public typealias SetupCGMCompletion = (CGMManager?) -> Void
    public func maybeSetupCGMManager(_ identifier: String, setupClosure: (CGMManagerUI.Type, @escaping SetupCGMCompletion) -> Void) {
        if identifier == pumpManager?.managerIdentifier, let cgmManager = pumpManager as? CGMManager {
            // We have a pump that is a CGM!
            self.cgmManager = cgmManager
        } else if let cgmManagerType = cgmManagerTypeByIdentifier(identifier) {
            setupClosure(cgmManagerType) {
                self.cgmManager = $0
            }
        }
    }
```
To me, this is clear: if we already have a `CGMManager`, assign it and move on.  If not, call a passed-in closure to do setup for me, and when done, assign it.

The thing that gets in the way is the `CGMManagerSetupViewControllerDelegate` which makes this dance awkward.  For that, I would introduce a shim.  Something like this:
```swift
private class DelegateShim: CGMManagerSetupViewControllerDelegate {
    let completion: (CGMManager?) -> Void
    init(completion: @escaping (CGMManager?) -> Void) {
        self.completion = completion
    }
    func cgmManagerSetupViewController(_ cgmManagerSetupViewController: CGMManagerSetupViewController, didSetUpCGMManager cgmManager: CGMManagerUI) {
        self.completion(cgmManager)
    }
}
```

Then, a caller might look like this:
```swift
        deviceManager.maybeSetupCGMManager(identifier) { cgmManagerType, setupCompletion in
            if var setupViewController = cgmManagerType.setupViewController(glucoseTintColor: .glucoseTintColor, guidanceColors: .default) {
                let shim = DelegateShim { setupCompletion($0) }
                setupViewController.setupDelegate = shim
                setupViewController.completionDelegate = self
                present(setupViewController, animated: true, completion: nil)
            } else {
                setupCompletion(cgmManagerType.init(rawState: [:]))
            }
        }
```

WDYT? (I've posted a commit with these diffs...)

**Pull request merged**
